runtime/unix.js

```diff
@@ -64,3 +64,180 @@ function unix_isatty(fileDescriptor) {
+    'EOPNOTSUPP', 'EPFNOSUPPORT', 'EAFNOSUPPORT', 'EADDRINUSE', 'EADDRNOTAVAIL',
+    'ENETDOWN', 'ENETUNREACH', 'ENETRESET', 'ECONNABORTED', 'ECONNRESET', 'ENOBUFS',
+    'EISCONN', 'ENOTCONN', 'ESHUTDOWN', 'ETOOMANYREFS', 'ETIMEDOUT', 'ECONNREFUSED',
+    'EHOSTDOWN', 'EHOSTUNREACH', 'ELOOP', 'EOVERFLOW'
+  ];
+  var variant = errors.indexOf(err.code);
+  if (variant < 0) {
+    // TODO: Make sure this works `EUNKNOWNERR(int)`
+    variant = BLOCK(0, err.errno);
+  }
+  var args = [
+    variant,
+    caml_string_of_jsbytes(err.syscall || ''),
+    caml_string_of_jsbytes(err.path || '')
+  ];
+  caml_raise_with_args(caml_named_value('Unix.Unix_error'), args.length, args);
+}
+
+//Provides: caml_stats_from_js
+function caml_stats_from_js(js_stats) {
+  /* ===Unix.file_kind===
+   * type file_kind =
+   *     S_REG                       (** Regular file *)
+   *   | S_DIR                       (** Directory *)
+   *   | S_CHR                       (** Character device *)
+   *   | S_BLK                       (** Block device *)
+   *   | S_LNK                       (** Symbolic link *)
+   *   | S_FIFO                      (** Named pipe *)
+   *   | S_SOCK                      (** Socket *)
+   */
+  var file_kind;
+  if (js_stats.isFile()) {
+    file_kind = 0;
+  }
+  if (js_stats.isDirectory()) {
+    file_kind = 1;
+  }
+  if (js_stats.isCharacterDevice()) {
+    file_kind = 2;
+  }
+  if (js_stats.isBlockDevice()) {
+    file_kind = 3;
+  }
+  if (js_stats.isSymbolicLink()) {
+    file_kind = 4;
+  }
+  if (js_stats.isFIFO()) {
+    file_kind = 5;
+  }
+  if (js_stats.isSocket()) {
+    file_kind = 6;
+  }
+  /* ===Unix.stats===
+   * type stats =
+   *  { st_dev : int;               (** Device number *)
+   *    st_ino : int;               (** Inode number *)
+   *    st_kind : file_kind;        (** Kind of the file *)
+   *    st_perm : file_perm;        (** Access rights *)
+   *    st_nlink : int;             (** Number of links *)
+   *    st_uid : int;               (** User id of the owner *)
+   *    st_gid : int;               (** Group ID of the file's group *)
+   *    st_rdev : int;              (** Device ID (if special file) *)
+   *    st_size : int;              (** Size in bytes *)
+   *    st_atime : float;           (** Last access time *)
+   *    st_mtime : float;           (** Last modification time *)
+   *    st_ctime : float;           (** Last status change time *)
+   *  }
+   */
+  return BLOCK(
+    0,
+    js_stats.dev,
+    js_stats.ino,
+    file_kind,
+    js_stats.mode,
+    js_stats.nlink,
+    js_stats.uid,
+    js_stats.gid,
+    js_stats.rdev,
+    js_stats.size,
+    js_stats.atimeMs,
+    js_stats.mtimeMs,
+    js_stats.ctimeMs
+  );
+}
+
+//Provides: unix_stat
+//Requires: fs_node_supported, caml_stats_from_js
+//Requires: caml_jsbytes_of_string
+//Requires: caml_raise_unix_exn_of_nodejs_error, caml_failwith
+function unix_stat(filename) {
```

Comment:
Theses functions bypass the filesystem proxy of jsoo. I wonder if this is the right. Let me think about this a bit more

Reply:
Refactored!